dev-tools/smoke.js

```diff
@@ -0,0 +1,435 @@
+			items: new Pool(onlyValid(dex.gen, dex.data.Items, i => dex.getItem(i)), this.prng),
+			abilities: new Pool(onlyValid(dex.gen, dex.data.Abilities, a => dex.getAbility(a)), this.prng),
+			moves: new Pool(onlyValid(dex.gen, dex.data.Movedex, m => dex.getMove(m),
+				m => (m !== 'struggle' && (m === 'hiddenpower' || m.substr(0, 11) !== 'hiddenpower'))), this.prng),
+		};
+	}
+
+	logProgress(p) {
+		// `\r` = return to the beginning of the line
+		// `\x1b[k` (`\e[K`) = clear all characters from cursor position to EOL
+		if (this.games) process.stdout.write('\r\x1b[K');
+		// Deliberately don't print a `\n` character so that we can overwrite
+		process.stdout.write(
+			`[${this.format}] P:${p.pokemon} I:${p.items} A:${p.abilities} M:${p.moves} = ${this.games}`);
+	}
+}
+
+module.exports = {TeamGenerator, Pool, CoordinatedPlayerAI, SmokeRunner};
+
+function onlyValid(gen, obj, getter, additional, nonStandard) {
+	return Object.keys(obj).filter(k => {
+		const v = getter(k);
+		return v.gen <= gen &&
+			(!v.isNonstandard || !!nonStandard) &&
+			(!additional || additional(k, v));
+	});
+}
+
+function getSignatures(dex, pools) {
+	const signatures = new Map();
+	for (const id of pools.items.possible) {
+		const item = dex.data.Items[id];
+		if (item.megaEvolves) {
+			const pokemon = toId(item.megaEvolves);
+			const combo = {item: id};
+			let combos = signatures.get(pokemon);
+			if (!combos) {
+				combos = [];
+				signatures.set(pokemon, combos);
+			}
+			combos.push(combo);
+		} else if (item.zMoveUser) {
+			for (const user of item.zMoveUser) {
+				const pokemon = toId(user);
+				const combo = {item: id};
+				if (item.zMoveFrom) combo.move = toId(item.zMoveFrom);
+				let combos = signatures.get(pokemon);
+				if (!combos) {
+					combos = [];
+					signatures.set(pokemon, combos);
+				}
+				combos.push(combo);
+			}
+		}
+	}
+	return signatures;
+}
+
+if (require.main === module) {
+	Dex.includeModData();
+	RejectionTracker.register();
+
+	// TODO: Add triple battles once supported by the AI.
+	const FORMATS = [
+		'gen7customgame', 'gen7doublescustomgame',
+		'gen6customgame', 'gen6doublescustomgame',
+		'gen5customgame', 'gen5doublescustomgame',
+		'gen4customgame', 'gen4doublescustomgame',
+		'gen3customgame', 'gen3doublescustomgame',
+		'gen2customgame',
+		'gen1customgame'];
+
+	// Because of our handwavy accounting of 'usage', as well as to account for ordering effects and
+	// the interplay of various combinations, we should actually run multiple cycles and exhaust our
+	// pools multiple times to provide more confidence that we're actually smoking out issues.
+	//
+	// If cycles is negative the smoke test will instead run indefinitely in case you wish to
+	// dedicate some computer resources to finding PS! edge case crashes instead of mining crypto,
+	// folding proteins, or searching for extraterrestrials/large primes.
```

Comment:
Does folding@home even matter ever since AlphaFold annihilated everyone else?

Reply:
Ahah reviewing my jokes, I love it!